File: frontend/src/api/missions.ts
import client from "./client";
import type { Mission, MissionCreate } from "../types/mission";

export async function fetchMissions(): Promise<Mission[]> {
  const { data } = await client.get<Mission[]>("/missions");
  return data;
}

export async function createMission(payload: MissionCreate): Promise<Mission> {
  const { data } = await client.post<Mission>("/missions", payload);
  return data;
}

export async function startMission(id: string): Promise<Mission> {
  const { data } = await client.post<Mission>(`/missions/${id}/start`);
  return data;
}

export async function stopMission(id: string): Promise<Mission> {
  const { data } = await client.post<Mission>(`/missions/${id}/stop`);
  return data;
}

File: backend/routers/missions.py
from fastapi import APIRouter, Request, HTTPException

from schemas.mission import Mission, MissionCreate, MissionStatus

router = APIRouter()


@router.post("", response_model=Mission)
async def create_mission(data: MissionCreate, request: Request):
    store = request.app.state.mission_store
    return store.create(data)


@router.get("", response_model=list[Mission])
async def list_missions(request: Request):
    store = request.app.state.mission_store
    return store.list_all()


@router.get("/{mission_id}", response_model=Mission)
async def get_mission(mission_id: str, request: Request):
    store = request.app.state.mission_store
    mission = store.get(mission_id)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")
    return mission


@router.post("/{mission_id}/start", response_model=Mission)
async def start_mission(mission_id: str, request: Request):
    store = request.app.state.mission_store
    drone = request.app.state.drone_service
    mission = store.update_status(mission_id, MissionStatus.ACTIVE)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")
    drone.start_mission()
    return mission


@router.post("/{mission_id}/stop", response_model=Mission)
async def stop_mission(mission_id: str, request: Request):
    store = request.app.state.mission_store
    drone = request.app.state.drone_service
    mission = store.update_status(mission_id, MissionStatus.COMPLETED)
    if not mission:
        raise HTTPException(status_code=404, detail="Mission not found")
    drone.stop_mission()
    return mission
